Add method to determine if enforcement is disabled to PathConfigMethod

michalvavrik opened the pull request:
closes: #17381

Provide a way to determine whether enforcement has been disabled when path config path ends with a wildcard (non-static path) and no path contains pattern. Main difference between newly exposed method and `matches` method is that https://github.com/keycloak/keycloak/blob/main/adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/authorization/PolicyEnforcer.java#L276 is avoided.

pedroigor reviewed (commented):
@michalvavrik Thanks for creating it here.

I'm a bit lost now about what we are trying to achieve here. After reading https://github.com/quarkusio/quarkus/issues/18818, I understand we want to avoid remote calls to Keycloak when the path is disabled.

Can you please confirm my understanding and how these changes are helping with it?

michalvavrik commented:
Hey @pedroigor ,

maybe some explanative JavaDoc is needed, but it can wait after your review. Please consider following flow:

// unless already cached
1. I need to determine whether for given request path, enforcement is disabled, done here https://github.com/keycloak/keycloak/blob/main/adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/authorization/PolicyEnforcer.java#L271
2. Path config path contains a wildcard and caching is enabled, therefore this call happens https://github.com/keycloak/keycloak/blob/main/adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/authorization/PolicyEnforcer.java#L276
3. than enforcement is determined from https://github.com/keycloak/keycloak/blob/main/adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/authorization/PolicyEnforcer.java#L293 or https://github.com/keycloak/keycloak/blob/main/adapters/oidc/adapter-core/src/main/java/org/keycloak/adapters/authorization/PolicyEnforcer.java#L286, therefore remote call has no relevance as far as enforcement is discussed (unless the path contains pattern)

So the idea is to spare remote requests when possible.

michalvavrik committed:
Add method to determine if enforcement is disabled to PathConfigMethod

closes: #17381

Newly exposed method `isEnforcementDisabled` on `PathConfigMatcher` tries to determine if enforcement is disabled locally if possible (path do not contain pattern). That is important when path contains wildcard and by default, Keycloak is called.

michalvavrik commented:
@pedroigor resolved conflicts, but did't have time for other adjustments, let's have a look if this is okay or suggest what needs to be done, thanks

pedroigor commented on authz/policy-enforcer/src/main/java/org/keycloak/adapters/authorization/PathConfigMatcher.java:
```diff
@@ -65,6 +65,20 @@ public class PathConfigMatcher extends PathMatcher<PathConfig> {
+    public boolean isEnforcementDisabled(String targetUri) {
+        PathConfig pathConfig = pathCache.get(targetUri);
+
+        if (pathConfig == null && !pathCache.containsKey(targetUri)) {
```
```suggestion
        if (pathConfig == null) {
```

michalvavrik replied:
`pathCache.containsKey(targetUri)` can be true even when `pathConfig == null` because even cases when path config wasn't found (`matches` does same check, `org.keycloak.adapters.authorization.PathCache#put`). Just by reading code, I guess this situation may happen when even any path config isn't matched (that's what code allows, not sure if application logic too).

michalvavrik commented:
Hey @pedroigor, this is not well communicated. if there are required changes, please tell me and I will do them. Same goes if the solution is not right in general (I mean, I can try to find other :-) no promise), but this long waiting for such simplistic PR is demotivating in regards to future contributions.

michalvavrik commented:
I'll close this as 4 months seems enough for review.
